### PR TITLE
[Migrated] - Generate Composite Images Redistributable

### PR DESCRIPTION
**Pull request by ivdiazsa**
This PR is a continuation of #42477. We need to have new artifacts published as part of the validation process, and this functionality is not enabled for builds from forked repositories, as explained here: https://github.com/dotnet/aspnetcore/pull/42477#issuecomment-1170614946

### CODE REVIEW

**Comment by dougbu** [comment]
My apologies @ivdiazsa. The new PR is more complete but insufficient. You're getting NuGet packages uploaded in artifacts now. But archives aren't being made available.

The way to get `BlobArtifacts` created is apparently limited to internal non-PR builds. If you can push your branch to the AzDO mirror repo and start a manual build in the [aspnetcore-ci-official](https://dev.azure.com/dnceng/internal/_build?definitionId=21) using it, that'll work. If not, let me know&hellip;

**Comment by ivdiazsa** [comment]
Hi @dougbu! I've been looking at this "Build Analysis" failure but can't make sense of it. Why is this physical path thing being "wrong" when I didn't touch anything like that? Also, please let me know if you have any feedback to address regarding my latest changes with the layout.

**Comment by dougbu** [comment]
@ivdiazsa I'm not sure how you could have caused these issues but let's leave the commits alone and redo the build to see if anything remains consistent. This will pick up any fixes done in 'main' over the last couple of days (anything after @Seeker186 started the most recent build).

**Comment by dougbu** [comment]
Looks good❕ Testing new archive creation in [build #20220715.3](https://dev.azure.com/dnceng/internal/_build/results?buildId=1883705&view=results).

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 1]
```diff
@@ -574,12 +579,44 @@ This package is an internal implementation of the .NET Core SDK and is not meant
...
+    </PropertyGroup>
+
+    <ItemGroup>
+      <RedistLayoutFiles Include="$(RedistSharedFrameworkLayoutRoot)**\*.*" />
+      <CompositeFiles Include="$(CompositeTargetDir)**\*.*" />
+    </ItemGroup>
+
+    <RemoveDir Directories="$(CompositeFrameworkLayoutRoot)" />
+    <MakeDir Directories="$(CompositeFrameworkLayoutRoot)" />
+    <Copy SourceFiles="@(RedistLayoutFiles)" DestinationFolder="$(CompositeFrameworkLayoutRoot)\%(RecursiveDir)" />
```
Comment (dougbu) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:603]: Why is this necessary given `$(CompositeTargetDir)` should contain versions of the same assemblies❔

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 2]
```diff
@@ -574,12 +579,44 @@ This package is an internal implementation of the .NET Core SDK and is not meant
...
+
+    <ItemGroup>
+      <AspnetFiles Include="$(LayoutAspnetPath)**\*.*" />
+      <AspnetFilesByName Include="@(AspnetFiles->'%(Filename)%(Extension)')" />
+      <CompositeFilesByName Include="@(CompositeFiles->'%(Filename)%(Extension)')" />
+      <FrameworkCompositesByName Include="@(CompositeFilesByName)" Exclude="@(AspnetFilesByName)" />
```
Comment (dougbu) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: full-composite.r2r.dll ends up in this item group. Need to manually include the file in `@(AspNetFiles)` or `@(AspNetFilesByName)`.

**Review by dougbu — COMMENTED** [review]
Need to correct where full-composite.r2r.dll is placed in the composite layout. https://dev.azure.com/dnceng/_apis/resources/Containers/11142022/BlobArtifacts?itemPath=BlobArtifacts%2Faspnetcore-runtime-7.0.0-rc.1.22365.3-linux-arm-composite.tar.gz for example looks wrong.

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj (hunk 1 above)
Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:603]: OIC, this is about the files that end up higher in the layout than Microsoft.NETCore.App and Microsoft.AspNetCore.App. Never mind.

Comment (dougbu) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:603]: Not a really big deal but the `%(DestinationFolder)` metadata ends up w/ two slashes at the end when `%(RecursiveDir)` is empty i.e. the files are placed directly in `$(CompositeFrameworkLayoutRoot)`. Suggest
```suggestion
    <Copy SourceFiles="@(RedistLayoutFiles)" DestinationFolder="$(CompositeFrameworkLayoutRoot)%(RecursiveDir)" />
```

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj (hunk 2 above)
Comment (ivdiazsa) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: I think this is good. I downloaded the tar.gz you linked above, and `full-composite.r2r.dll` is in the NETCore.App subtree. That's how all our tests were done when we did fully manual builds to know what to code in the actual repo.

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: What tests are we talking about❔

It seems very strange that the composite image could work at all in either directory because IL would be missing for either AspNetCore or NETCore assemblies. I'm thinking here of https://github.com/dotnet/aspnetcore/pull/41855#issuecomment-1159784470

Comment (ivdiazsa) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: The reason this layout works well is that the composite is looked for in the same folder _System.Private.CoreLib.dll_, which is the first loaded assembly, is located in. I validated this by using a _.NET SDK Docker Image_. I created a simple ASP.NET web app, and ensured it ran with the untouched container. Then, I replaced the corresponding parts of the dotnet installation with the composite product (_CompositeFx.Layout_) built in my clone of the ASP.NET repo. Then, ran the simple web app again, and it worked. @davidwrighton and @trylek have also validated that this layout works because of the aforementioned reason.

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: I don't believe this layout actually works for customers because
1. it's false advertising, this composite image is _not_ reflective of what will land in the .dotnet/shared/Microsoft.NETCore.App/ folder during a regular local (dotnet-install.sh) installation
2. the name is the same as the dotnet/runtime composite image and whichever of the runtime installers and ASP.NET archives is installed second will "win" -- assuming the archives will eventually be available for dotnet-install.sh installations or on <https://dotnet.microsoft.com/en-us/download/dotnet>.
3. the ASP.NET composite assembly likely can't be included side-by-side w/ the runtime composite assembly if we want to put both into our regular .NET installers

Suggest this be treated as a bug w/ either the composite assembly name in dotnet/runtime or in the System.Private.CoreLib.dll discovery code. (From your experience, are you sure the composite assembly is used for ASP.NET code and not just the lower-level runtime code❔)

Comment (mangod9) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: Hi @dougbu, agreed that we need to have cohesive strategy here, for now we just would like to get a full composite built and included in a asp.net container image. The goal is not to ship composite images with OOB experience at the moment. The main goal is to ensure for the container scenario we can enable the 20% or so startup gain we hope composites will enable. 

Adding @richlander too. Let me know if a quick sync would help to move this forward.

Comment (mangod9) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: Perhaps renaming the composite would be a short term workaround?

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: I would appreciate a sync because I'm not understanding the overall requirements and restrictions. I _thought_ I got it but, nooo&hellip;

In the meantime, _which_ composite are you suggesting we rename @mangod9❔

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 3]
```diff
@@ -467,15 +467,15 @@ This package is an internal implementation of the .NET Core SDK and is not meant
   </Target>
 
   <Target Name="_GenerateComposites"
-    DependsOnTargets="_ExpandRuntimePackageRoot">
+          Condition="'$(TargetOsName)' == 'linux'"
```
Comment (wtgodbe) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:470]: Do we need this on linux-musl? If so switch this to `BuildOsName`

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 4]
```diff
@@ -657,14 +657,10 @@ This package is an internal implementation of the .NET Core SDK and is not meant
   </Target>
 
   <Target Name="_CreateRedistCompositeArchive"
+          Condition="'$(TargetOsName)' == 'linux'"
```
Comment (wtgodbe) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:660]: Same here

**Review by wtgodbe — APPROVED** [review]

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 5]
```diff
@@ -618,12 +656,28 @@ This package is an internal implementation of the .NET Core SDK and is not meant
...
+    <Message Importance="High" Text="$(MSbuildProjectFile) -> $(CompositeArchiveOutputPath)" />
+    <ZipDirectory
+      SourceDirectory="$(CompositeFrameworkLayoutRoot)"
+      DestinationFile="$(CompositeArchiveOutputPath)"
+      Overwrite="true"
+      Condition="'$(ArchiveExtension)' == '.zip'" />
+    <Exec
+      Command="tar -czf $(CompositeArchiveOutputPath) ."
+      WorkingDirectory="$(CompositeFrameworkLayoutRoot)"
+      Condition="'$(ArchiveExtension)' == '.tar.gz'" />
```
[the remark this answers is not in the document]
Comment (ivdiazsa) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:667]: I was sure something had eluded me but didn't know where. Thanks for pointing it out Will

**Review by wtgodbe — APPROVED** [review]

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 6]
```diff
@@ -574,12 +579,44 @@ This package is an internal implementation of the .NET Core SDK and is not meant
...
           DestinationFiles="@(SharedFxContent->'$(RedistLayoutTargetDir)%(FileName)%(Extension)')"
           UseHardlinksIfPossible="true" />
     <Message Importance="High" Text="$(MSbuildProjectFile) -> $(RedistLayoutTargetDir)" />
   </Target>
 
+  <Target Name="_BatchCopyToCompositeLayout"
+          Condition="'$(TargetOsName)' == 'linux'">
+    <PropertyGroup>
+      <CompositeFrameworkLayoutRoot>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', CompositeFx.Layout, '$(Configuration)', '$(TargetRuntimeIdentifier)'))</CompositeFrameworkLayoutRoot>
+      <LayoutAspnetPath>$(CompositeFrameworkLayoutRoot)\shared\$(SharedFxName)\$(SharedFxVersion)\</LayoutAspnetPath>
```
Comment (dougbu) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:593]: This may be irrelevant but how is this part of the layout used❔ That is, will these assemblies be found when using the composite assembly in the Microsoft.NETCore.App part of the layout❔

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj (hunk 2 above)
Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: > The reason this layout works well is that the composite is looked for in the same folder System.Private.CoreLib.dll, which is the first loaded assembly, is located in. I validated this by using a .NET SDK Docker Image. I created a simple ASP.NET web app, and ensured it ran with the untouched container. Then, I replaced the corresponding parts of the dotnet installation with the composite product (CompositeFx.Layout) built in my clone of the ASP.NET repo. Then, ran the simple web app again, and it worked. @davidwrighton and @trylek have also validated that this layout works because of the aforementioned reason.

Thanks for getting this in. I have a followup question based on this earlier answer: I get how NETCore assemblies and their IL are loaded. But how can the composite find the AspNetCore assemblies (that are in a different folder from the composite) w/ the current archive layout❔

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: If I'm right, I lean toward putting all of the non-fxr assemblies in a single shared/ folder, preferably shared/Microsoft.AspNetCore.App (because the composite is "for" ASP.NET Core unless `dotnet` knows to find System.Private.CoreLib.dll in shared/Microsoft.NETCore.App.

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: > there is only one composite,

This isn't exactly true though I think the dotnet/runtime composite was renamed since I last looked. It's now named Microsoft.NETCore.App.Composite.r2r.dll and is bundled in archives like dotnet-runtime-composite-7.0.0-rc.1.22378.4-linux-arm64.tar.gz (which contains just one shared/ folder). The ASP.NET Core composite is named full-composite.r2r.dll and bundled in archives like aspnetcore-runtime-7.0.0-rc.1.22378.3-linux-arm64-composite.tar.gz (which contains two shared/ folders).

1. Are both of those archives in use❔
2. If no, could the unused one be removed❔
3. If yes,
  1. Should we make the archive names match i.e., change to name the ASP.NET Core archive names to something like aspnetcore-runtime-composite-7.0.0-rc.1.22378.3-linux-arm64.tar.gz❔
  2. Should we make the assembly names more similar i.e., change the ASP.NET Core composite name to Microsoft.AspNetCore.App.Composite.r2r.dll❔
  3. Should we include the Microsoft.NETCore.App.Composite.r2r.dll assembly (plus any assemblies from `dotnet-runtime-composite-*` archives we aren't already including in `dotnet-aspnetcore-*-composite` archives) in the ASP.NET Core archive❔

Comment (mangod9) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: the .NET app composite is experimental as well, its not being shipped/consumed downstream. We ran experiments where we tested separate Fx and asp.net composites with a (Fx+asp.net) composite and the latter had better performance as might be expected due to better inlining across modules.

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: That helps @mangod9 but I'm still thinking IL for the assemblies in the shared/Microsoft.AspNetCore.App/ folder aren't currently useful because they won't be found.

Comment (ivdiazsa) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: The ASP.NET binaries are indeed found and used by the composite. If they weren't, we would be able to just copy the NETCore.App binaries, and that's not the case. Doing so makes dotnet fail with an MVID mismatch error, which rightfully says the Aspnet binaries present are not the ones the composite is looking for.

`Process terminated. MVID mismatch between loaded assembly 'Microsoft.AspNetCore' (MVID = {ffe9fdbb-049b-4048-933b-f643f09f799f}) and an assembly with the same simple name embedded in the native image 'full-composite.r2r.dll' (MVID = {a20c1aa2-442f-4d51-9e0a-48c11b8b8767})
   at Program.<Main>$(System.String[])
Aborted`

Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:609]: To make sure I'm understanding: If I were to expand a `dotnet-aspnetcore-*-composite` archive (w/o my proposed renaming) in a `dotnet` installation directory, then delete the new `shared/Microsoft.AspNetCore.App/{version}/` folder, subsequent application loads using the shared runtime (well, those using the specific runtime versions) would fail❔
